### PR TITLE
feat: Parse registry config from .yarnrc.yml

### LINKED ISSUE #16353: Parse registryUrls from .yarnrc.yml
### What would you like Renovate to be able to do?

Parse `registryURls` from `.yarnrc.yml` (for Yarn 2/3)

### If you have any ideas on how this should be implemented, please tell us here.

Reproduction: https://github.com/renovate-reproductions/renovate-private-registry-yarn-v3-bug

Specifically the `npmRegistryServer` lines should be noted: https://github.com/renovate-reproductions/renovate-private-registry-yarn-v3-bug/blob/6ca4b4e59640ace4df1d95ae2760a4e8d95ac357/.yarnrc.yml#L7

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by avaly**
## Changes

Parse registry configuration from `.yarnrc.yml`.

## Context

Closes #16353

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — CHANGES_REQUESTED** [review]
don't like the idea to use npmrc, instead hostrules and registry url on dependency should be used

**Comment by avaly** [comment]
@viceice I've started to convert the code to parse the Yarn config into `NpmrcRules` (with `hostRules` and `packageRules`, but I don't see how I can use that result type in the [`extractPackageFile`](https://github.com/avaly/renovate/blob/main/lib/modules/manager/npm/extract/index.ts#L38-L42) method which needs to return [`PackageFile`](https://github.com/avaly/renovate/blob/main/lib/modules/manager/types.ts#L66), which does not include `hostRules` or `packageRules`. Any pointers?

**Comment by viceice** [comment]
`registryUrls`. is a direct property, so jo need for package rule.

the auth thing depends on another feature, which allows configure auth via package rule, so a manager could also extract it and directly add to result.

@rarkins i think we should add a optional prepare function to managers, which parses those config files and returns host rules and package rule.

so we only read and parse that file once.

WDYT?

**Comment by rarkins** [comment]
Wouldn't any such prepare function need to run on the same `fileMatch` contents anyway? I'm not sure I see any benefit if so.

Can we just add `registryUrls` to both the package file and dependency interfaces?

File: lib/modules/manager/npm/extract/index.ts [hunk 1]
```diff
@@ -139,6 +144,19 @@ export async function extractPackageFile(
...
+    if (is.string(config.npmrc) && !config.npmrcMerge) {
+      logger.debug(
+        { yarnrcYmlFileName },
+        'Repo .yarnrc.yml file is ignored due to config.npmrc with config.npmrcMerge=false'
+      );
+    } else {
+      yarnRules = convertYarnrcYmlToRules(repoYarnrcYml);
+    }
+  }
+
```
Comment (fgreinacher) [inline lib/modules/manager/npm/extract/index.ts:159]: Would it maybe easier if we just transform the `yarnrc` to `npmrc`? That way we would not need additional logic for mapping the rules to renovate config.

**Comment by fgreinacher** [comment]
@avaly Are you still able to work on this? I'd be happy to help get this in.

@rarkins I'd also like this to support legacy `.yarnrc` files. Do you prefer a separate issue/MR or is it fine if we add it here?

**Comment by avaly** [comment]
@fgreinacher I won't have availability to work on this soon. Feel free to take over the code and add the missing parts.

File: lib/modules/manager/npm/extract/index.ts (hunk 1 above)
Comment (fgreinacher) [reply lib/modules/manager/npm/extract/index.ts:159]: @viceice I just noticed that it was like that before and [you didn't like](https://github.com/renovatebot/renovate/pull/16413#pullrequestreview-1027586525) it. Can you explain a bit why that's the case? Doing the transformation once feels much less invasive than what we have right now.

Comment (rarkins) [reply lib/modules/manager/npm/extract/index.ts:159]: We're trying to reduce our "dependency" on `npmrc`. So by transforming to it here we may make it harder to improve our `npmrc` approach

**Comment by avaly** [comment]
This seems to have been implemented in #19864

**Pull request closed without merging**